src/crimson/osd/object_context_loader.cc

```diff
@@ -105,28 +105,43 @@ using crimson::common::local_conf;
     if constexpr (track) {
       obc->append_to(obc_set_accessing);
     }
-    return obc->with_lock<State, IOInterruptCondition>(
-      [existed=existed, obc=obc, func=std::move(func), this]() mutable {
-      return get_or_load_obc<State>(
-	obc, existed
-      ).safe_then_interruptible(std::move(func));
-    }).finally([FNAME, this, obc=std::move(obc)] {
-      DEBUGDPP("released object {}", dpp, obc->get_oid());
-      if constexpr (track) {
-	obc->remove_from(obc_set_accessing);
-      }
-    });
+    if (existed) {
+      return obc->with_lock<State, IOInterruptCondition>(
+	[func=std::move(func), obc=ObjectContextRef(obc)] {
+	  return std::invoke(std::move(func), obc);
+	}
+      ).finally([FNAME, this, obc=ObjectContextRef(obc)] {
+	DEBUGDPP("released object {}", dpp, obc->get_oid());
+	if constexpr (track) {
+	  obc->remove_from(obc_set_accessing);
+	}
+      });
+    } else {
+      return obc->load_then_with_lock<State> (
```

Comment:
Ack. The _first_ user will get the exclusive lock in construction atomically which will cause the next users (`existed=true`) to await in `obc->with_lock` until loading is finished (aka demoted).

Reply:
Right.  Whoever places the obc in the registry to begin with will, without waiting for a future, obtain an exclusive lock.  It will then hold that exclusive lock until the obc is loaded and then demote to the originally needed lock.